src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -241,6 +280,240 @@ void DiscoveryDataBase::AckedFunctor::operator () (
+        }
+    }
+
+    for (auto rit = readers_.begin(); rit != readers_.end(); ++rit)
+    {
+        if (rit->first.guidPrefix == participant_guid.guidPrefix)
+        {
+            readers_.erase(rit->first);
+            --rit;
+        }
+    }
+
+    // Delete Participant entries from readers_by_topic and writers_by_topic
+    for (auto tit = readers_by_topic_.begin(); tit != readers_by_topic_.end(); ++tit)
+    {
+        for (auto rit = tit->second.begin(); rit != tit->second.end(); ++rit)
+        {
+            if(rit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(rit);
+                --rit;
+            }
+        }
+    }
+
+    for (auto tit = writers_by_topic_.begin(); tit != writers_by_topic_.end(); ++tit)
+    {
+        for (auto wit = tit->second.begin(); wit != tit->second.end(); ++wit)
+        {
+            if(wit->guidPrefix == participant_guid.guidPrefix)
+            {
+                tit->second.erase(wit);
+                --wit;
+            }
+        }
+    }
+
+    // Remove participant from others participants_[]::relevant_participants_builtin_ack_status
+    for (auto pit = participants_.begin(); pit != participants_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others writers_[]::relevant_participants_builtin_ack_status
+    for (auto wit = writers_.begin(); wit != writers_.end(); ++wit)
+    {
+        wit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    // Remove participant from others readers_[]::relevant_participants_builtin_ack_status
+    for (auto pit = readers_.begin(); pit != readers_.end(); ++pit)
+    {
+        pit->second.remove_participant(participant_guid.guidPrefix);
+    }
+
+    if(std::find(disposals_.begin(), disposals_.end(), ch) == disposals_.end())
+    {
+        disposals_.push_back(ch);
+    }
+}
+
+void DiscoveryDataBase::create_participant_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    DiscoveryParticipantInfo part(ch);
+    participants_.insert(std::make_pair(guid_from_change(ch).guidPrefix, part));
+}
+
+void DiscoveryDataBase::create_writers_from_change(
+        eprosima::fastrtps::rtps::CacheChange_t* ch,
+        const eprosima::fastrtps::string_255& topic_name)
+{
+    const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
+
+    DiscoveryEndpointInfo tmp_writer(ch, topic_name);
+    std::pair<std::map<eprosima::fastrtps::rtps::GUID_t,DiscoveryEndpointInfo>::iterator,bool> ret;
+    ret = writers_.insert(std::make_pair(writer_guid, tmp_writer));
+
+    if (ret.second) {
+        auto readers_it = readers_by_topic_.find(topic_name);
+        if (readers_it != readers_by_topic_.end())
+        {
+            for (auto reader_it: readers_it->second)
+            {
+                writers_.find(writer_guid)->second.add_or_update_ack_participant(reader_it.guidPrefix);
```

Comment:
I think you can do this find before the loops and store the iterator. Then you just access to it inside the loops. That way you only look once, since the result is always going to be the same